crates/router/src/core/utils.rs

```diff
@@ -1254,3 +1315,79 @@ pub fn get_incremental_authorization_allowed_value(
         incremental_authorization_allowed
     }
 }
+
+pub(super) trait GetProfileId {
+    fn get_profile_id(&self) -> Option<&String>;
+}
+
+impl GetProfileId for MerchantConnectorAccount {
+    fn get_profile_id(&self) -> Option<&String> {
+        self.profile_id.as_ref()
+    }
+}
+
+impl GetProfileId for storage::PaymentIntent {
+    fn get_profile_id(&self) -> Option<&String> {
+        self.profile_id.as_ref()
+    }
+}
+
+impl<A> GetProfileId for (storage::PaymentIntent, A) {
+    fn get_profile_id(&self) -> Option<&String> {
+        self.0.get_profile_id()
+    }
+}
+
+impl GetProfileId for diesel_models::Dispute {
+    fn get_profile_id(&self) -> Option<&String> {
+        self.profile_id.as_ref()
+    }
+}
+
+#[cfg(feature = "payouts")]
+impl GetProfileId for PayoutData {
+    fn get_profile_id(&self) -> Option<&String> {
+        Some(&self.profile_id)
+    }
+}
+
+/// Filter Objects based on profile ids
+pub(super) fn filter_objects_based_on_profile_id_list<T: GetProfileId>(
+    profile_id_list_auth_layer: Option<Vec<String>>,
+    object_list: Vec<T>,
+) -> Vec<T> {
+    if let Some(profile_id_list) = profile_id_list_auth_layer {
+        let profile_ids_to_filter: HashSet<_> = profile_id_list.iter().collect();
+        object_list
+            .into_iter()
+            .filter_map(|item| {
+                if item
+                    .get_profile_id()
+                    .is_some_and(|profile_id| profile_ids_to_filter.contains(profile_id))
+                {
+                    Some(item)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    } else {
+        object_list
+    }
+}
+
+pub(super) fn validate_profile_id_from_auth_layer<T: GetProfileId>(
+    profile_id_auth_layer: Option<&String>,
+    object: &T,
+) -> RouterResult<()> {
+    if profile_id_auth_layer
+        .is_some_and(|auth_profile_id| Some(auth_profile_id) == object.get_profile_id())
```

Comment:
This will still break, if we don't find the profile_id in `T`. What are the cases where we will get profile_id as null?

If the profile_id is not supposed to be null, isn't it data integrity issue. And shouldn't we throw 500?

Reply:
If we can do this, then we make the trait function return the profile id directly instead of option.